src/pointer.rs

```diff
@@ -1,88 +1,48 @@
 use crate::{
-    Assignment, Error, MalformedPointerError, NotFoundError, ReplaceTokenError, Token, Tokens,
-    UnresolvableError,
+    assign::assign_value, parse_error, AssignError, Assignment, InvalidEncodingSnafu, ParseError,
+    ReplaceTokenError, Resolve, ResolveMut, Token, Tokens,
 };
 use alloc::{
-    borrow::{Cow, ToOwned},
+    borrow::ToOwned,
     string::{String, ToString},
     vec::Vec,
 };
 use core::{borrow::Borrow, cmp::Ordering, mem, ops::Deref, slice, str::FromStr};
 use serde::{Deserialize, Serialize};
-use serde_json::{map::Entry, Map, Value};
-
-/// Helper type for deserialization. Either a valid [`Pointer`] or a string and
-/// the parsing error
-#[derive(Debug, Clone, PartialEq, Eq)]
-pub enum MaybePointer {
-    /// A valid [`Pointer`].
-    Pointer(PointerBuf),
-    /// An invalid [`Pointer`] and the error that caused it.
-    Malformed(String, MalformedPointerError),
-}
-impl Deref for MaybePointer {
-    type Target = str;
+use serde_json::Value;
+use snafu::ensure;
 
-    fn deref(&self) -> &Self::Target {
-        match self {
-            MaybePointer::Pointer(p) => p.as_str(),
-            MaybePointer::Malformed(s, _) => s,
-        }
-    }
-}
-impl<'de> Deserialize<'de> for MaybePointer {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        String::deserialize(deserializer).map(|s| match PointerBuf::from_str(&s) {
-            Ok(ptr) => MaybePointer::Pointer(ptr),
-            Err(e) => MaybePointer::Malformed(s, e),
-        })
-    }
-}
-impl Serialize for MaybePointer {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: serde::Serializer,
-    {
-        match self {
-            MaybePointer::Pointer(ptr) => serializer.serialize_str(ptr.as_str()),
-            MaybePointer::Malformed(s, _) => serializer.serialize_str(s),
-        }
-    }
-}
-
-fn validate(value: &str) -> Result<&str, MalformedPointerError> {
+fn validate(value: &str) -> Result<&str, ParseError> {
     let mut chars = value.chars();
-
     match chars.next() {
         Some('#') => {
+            // we forgive the usage of `'#'` at the beginning of a pointer even
+            // though this is technically incorrect.
+            // it is a common enough mistake as json pointers are often used in
+            // fragments of URIs to indicate sub-resources.
```

Comment:
Should we take advantage of the major release to remove this relaxation of the spec?

Reply:
I'm cool with that. I'll remove it.